File: compiler/src/syntax/attributes/AttributeValue.swift
public struct StringAttributeValueSyntax: Sendable {
    public let span: Span
    public let value: Substring
}

func parseStringAttributeValue(with parser: Parser) throws(ParseError) -> StringAttributeValueSyntax
{
    let span = parser.spanned()
    let value = try parser.token(.string)
    return StringAttributeValueSyntax(span: span(), value: value)
}

public struct ConnectionAttributeValueSyntax: Sendable {
    public let span: Span
    public let left: Substring
    public let right: Substring
    public let label: Substring
}

func parseConnectionAttributeValue(with parser: Parser) throws(ParseError)
    -> ConnectionAttributeValueSyntax
{
    let span = parser.spanned()
    let left = try parser.token(.lowercaseName)
    try parser.token(.functionOperator)
    let right = try parser.token(.lowercaseName)
    try parser.token(.leftParenthesis)
    let label = try parser.token(.string)
    try parser.token(.rightParenthesis)

    return ConnectionAttributeValueSyntax(span: span(), left: left, right: right, label: label)
}

private let attributeValueCacheToken = Parser.CacheToken()

func parseAttributeValue(with parser: Parser) throws(ParseError) -> any Visitable {
    try parser.parseCached(token: attributeValueCacheToken) { () throws(ParseError) in
        if let value = try parser.parseOptional(parseStringAttributeValue) { return value }

        if let value = try parser.parseOptional(parseConnectionAttributeValue) { return value }

        throw parser.error("Expected an attribute value")
    }
}

extension StringAttributeValueSyntax: Visitable {
    public func visit(node: Node, with visitor: Visitor) {}
}

extension ConnectionAttributeValueSyntax: Visitable {
    public func visit(node: Node, with visitor: Visitor) {}
}

File: compiler/syntax/src/attributes/attribute_value.rs
use wipple_core::{
    ast::AstKey,
    visit::attributes::{ConnectionAttributeValue, StringAttributeValue},
};
use wipple_parse::{
    lexer::TokenKind,
    names::parse_type_parameter_name,
    parse_alt,
    parser::{ParseError, Parser},
};

pub fn parse_attribute_value(parser: &mut Parser<'_>) -> Result<AstKey, ParseError> {
    parse_alt!(parser, {
        parse_string_attribute_value as value => parser.in_ast(value),
        parse_connection_attribute_value as value => parser.in_ast(value),
        _ => "Expected an attribute value",
    })
}

pub fn parse_string_attribute_value(
    parser: &mut Parser<'_>,
) -> Result<StringAttributeValue, ParseError> {
    let span = parser.spanned();
    let value = parser.token(TokenKind::String)?;
    Ok(StringAttributeValue {
        span: span(parser),
        value,
    })
}

pub fn parse_connection_attribute_value(
    parser: &mut Parser<'_>,
) -> Result<ConnectionAttributeValue, ParseError> {
    let span = parser.spanned();
    let left = parse_type_parameter_name(parser)?;
    parser.token(TokenKind::FunctionOperator)?;
    let right = parse_type_parameter_name(parser)?;
    parser.token(TokenKind::LeftParenthesis)?;
    let label = parser.token(TokenKind::String)?;
    parser.token(TokenKind::RightParenthesis)?;

    Ok(ConnectionAttributeValue {
        span: span(parser),
        left,
        right,
        label,
    })
}
